transactor/src/main/scala/io/mediachain/copycat/S3BackingStore.scala

```diff
@@ -0,0 +1,248 @@
+import java.util.concurrent.{ExecutorService, Executors, TimeUnit}
+import java.io.{File, FileOutputStream}
+import com.amazonaws.AmazonClientException
+import com.amazonaws.auth.{AWSCredentials, BasicAWSCredentials}
+import com.amazonaws.services.s3.AmazonS3Client
+import org.slf4j.{Logger, LoggerFactory}
+import scala.concurrent.Await
+import scala.concurrent.duration.Duration
+import scala.collection.JavaConversions._
+import scala.util.{Try, Success, Failure}
+import cats.data.Xor
+import io.mediachain.protocol.Transactor._
+import io.mediachain.protocol.Datastore._
+import io.mediachain.protocol.CborSerialization._
+import io.mediachain.copycat.Client.{ClientState, ClientStateListener}
+import io.mediachain.datastore.DynamoDatastore
+
+class S3BackingStore(config: S3BackingStore.Config)
+extends JournalListener with ClientStateListener with AutoCloseable {
+  import S3BackingStore.multihashRef
+  private val logger = LoggerFactory.getLogger(classOf[S3BackingStore])
+  private val exec = Executors.newScheduledThreadPool(Math.max(4, Runtime.getRuntime.availableProcessors))
+  private val datastore = new DynamoDatastore(config.dynamo)
+  private val s3 = new AmazonS3Client()
+  private val s3bucket = config.s3bucket
+  private val client = Client.build(config.sslConfig)
+  private var cluster: Option[List[String]] = None
+  private var blocks: Set[Reference] = Set()
+  
+  client.addStateListener(this)
+
+  def connect(addrs: List[String]) {
+    cluster = Some(addrs)
+    client.connect(addrs)
+    client.listen(this)
+  }
+  
+  def close() {
+    exec.shutdownNow()
+    client.close()
+    datastore.close()
+    s3.shutdown()
+  }
+  
+  // Client state listener
+  def onStateChange(state: ClientState) {
+    state match {
+      case ClientState.Connected =>
+        logger.info("Client connected; scheduling blockchain catchup")
+        exec.submit(new Runnable {
+          def run { withErrorLog(writeCurrentBlock()) }
+        })
+      case ClientState.Suspended => ()
+      case ClientState.Disconnected =>
+        logger.info("Client disconnected; scheduling reconnect")
+        exec.submit(new Runnable {
+          def run { withErrorLog(reconnect()) }
+        })
+    }
+  }
+  
+  // JournalListener
+  def onJournalCommit(entry: JournalEntry) {}
+
+  def onJournalBlock(ref: Reference) {
+    logger.info(s"New block ${ref}")
+    execWriteBlock(ref)
+  }
+  
+  private def withErrorLog(expr: => Unit) {
+    try {
+      expr
+    } catch {
+      case e: Throwable =>
+        logger.error("Unhandled exception in task", e)
+    }
+  }
+  
+  private def writeCurrentBlock() {
+    logger.info("Fetching current block")
+    val op = client.currentBlock
+    val block = Await.result(op, Duration.Inf)
+    logger.info("Current blockchain pointer is " + block.chain)
+    block.chain.foreach { execWriteBlock(_) }
+  }
+  
+  private def execWriteBlock(ref: Reference) {
+    this.synchronized {
+      if (!blocks.contains(ref)) {
```

Comment:
It will walk up the chain until it finds a block it already has in the list of visited blocks or reaches nil.
It will not write a block a second time, because it will either be in the visited blocks set (which will make execWriteBlock a noop) or already in S3 (the wirte checks to see if it is already present in S3 with listObjects)

The reason for this implementation is that it allows the backing store to catch up in the block chain and be easily restartable.